src/Microsoft.ML.PCA/PcaTransform.cs

```diff
@@ -190,90 +281,104 @@ private static VersionInfo GetVersionInfo()
                 loaderAssemblyName: typeof(PcaTransform).Assembly.FullName);
         }
 
-        // These are parallel to Infos.
-        private readonly ColumnType[] _types;
+        private readonly int _numColumns;
+        private readonly ColumnInfo[] _columns;
         private readonly TransformInfo[] _transformInfos;
 
-        private readonly int[] _oversampling;
-        private readonly bool[] _center;
-        private readonly int[] _weightColumnIndex;
-
         private const string RegistrationName = "Pca";
 
         /// <summary>
         /// Public constructor corresponding to SignatureDataTransform.
         /// </summary>
-        public PcaTransform(IHostEnvironment env, Arguments args, IDataView input)
-            : base(env, RegistrationName, Contracts.CheckRef(args, nameof(args)).Column,
-                input, TestIsFloatItem)
+        public PcaTransform(IHostEnvironment env, IDataView input, ColumnInfo[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(PcaTransform)), GetColumnPairs(columns))
         {
-            Host.AssertNonEmpty(Infos);
-            Host.Assert(Infos.Length == Utils.Size(args.Column));
+            Host.AssertNonEmpty(ColumnPairs);
 
-            _transformInfos = new TransformInfo[args.Column.Length];
-            _oversampling = new int[args.Column.Length];
-            _center = new bool[args.Column.Length];
-            _weightColumnIndex = new int[args.Column.Length];
-            for (int i = 0; i < _transformInfos.Length; i++)
+            _columns = columns;
+            _numColumns = columns.Length;
+            _transformInfos = new TransformInfo[_numColumns];
+
+            for (int i = 0; i < _numColumns; i++)
             {
-                Host.Check(Infos[i].TypeSrc.VectorSize > 1, "Pca transform can only be applied to columns with known dimensionality greater than 1");
-                _transformInfos[i] = new TransformInfo(args.Column[i], args, Infos[i].TypeSrc.ValueCount);
-                _center[i] = args.Column[i].Center ?? args.Center;
-                _oversampling[i] = args.Column[i].Oversampling ?? args.Oversampling;
-                Host.CheckUserArg(_oversampling[i] >= 0, nameof(args.Oversampling), "Oversampling must be non-negative");
-                _weightColumnIndex[i] = -1;
-                var weightColumn = args.Column[i].WeightColumn ?? args.WeightColumn;
-                if (weightColumn != null)
-                {
-                    if (!Source.Schema.TryGetColumnIndex(weightColumn, out _weightColumnIndex[i]))
-                        throw Host.Except("weight column '{0}' does not exist", weightColumn);
-                    var type = Source.Schema.GetColumnType(_weightColumnIndex[i]);
-                    Host.CheckUserArg(type == NumberType.Float, nameof(args.WeightColumn));
-                }
+                var col = columns[i];
+                col.SetSchema(input.Schema);
```

Comment:
>SetSchema [](start = 20, length = 9)

Yea, that's bad. Let's not do it

Reply:
This is an internal functionality not visible to the user. Instead of having many many arrays in the transformer and mapper class with the input/output column type and index information, I'm bundling them all in the ColumnInfo, instead of creating another class with column information. What's the downside?

---
In reply to: [227070094](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227070094) [](ancestors = 227070094)